listen.go

```diff
@@ -31,10 +34,46 @@ func reuseUnixSocket(network, addr string) (any, error) {
 }
 
 func listenReusable(ctx context.Context, lnKey string, network, address string, config net.ListenConfig) (any, error) {
-	switch network {
-	case "udp", "udp4", "udp6", "unixgram":
+	var socketFile *os.File
+
+	fd := slices.Contains([]string{"fd", "fdgram"}, network)
+	if fd {
+		socketFd, err := strconv.ParseUint(address, 0, strconv.IntSize)
+		if err != nil {
+			return nil, fmt.Errorf("invalid file descriptor: %v", err)
+		}
+
+		func() {
```

Comment:
this is the only significantly changed behavior from https://github.com/caddyserver/caddy/pull/6543 ; `listenReusable` locks the fd used to to open files from fd numbers. this is similar to what `reuseUnixSocket` does, but it does not need to handle any messy race conditions involving unlink or the filesystem. it just needs to ensure that every fd is opened at most once, so the lock prevents multiple  reads before the first write.